File: solutions/atc001/dfs_a_naive.rb
# frozen_string_literal: true

DIRECTIONS = [
  [-1, 0],
  [1, 0],
  [0, -1],
  [0, 1]
].freeze

height, width = $stdin.gets.split.map(&:to_i)
grid = Array.new(height) { $stdin.gets.chomp.chars }

start = nil
goal = nil

grid.each_with_index do |row, y|
  row.each_with_index do |cell, x|
    start = [y, x] if cell == 's'
    goal = [y, x] if cell == 'g'
  end
end

queue = [start]
head = 0
visited = Array.new(height) { Array.new(width, false) }
visited[start[0]][start[1]] = true

reachable = false

while head < queue.length
  y, x = queue[head]
  head += 1

  if [y, x] == goal
    reachable = true
    break
  end

  DIRECTIONS.each do |dy, dx|
    ny = y + dy
    nx = x + dx

    next if ny.negative? || ny >= height
    next if nx.negative? || nx >= width
    next if grid[ny][nx] == '#'
    next if visited[ny][nx]

    visited[ny][nx] = true
    queue << [ny, nx]
  end
end

puts(reachable ? 'Yes' : 'No')

File: solutions/atc001/dfs_a_naive.py
from collections import deque
import sys


DIRECTIONS = [(-1, 0), (1, 0), (0, -1), (0, 1)]


def main() -> None:
    input = sys.stdin.readline
    height, width = map(int, input().split())
    grid = [list(input().strip()) for _ in range(height)]

    start = None
    goal = None

    for y, row in enumerate(grid):
        for x, cell in enumerate(row):
            if cell == "s":
                start = (y, x)
            elif cell == "g":
                goal = (y, x)

    queue = deque([start])
    visited = [[False] * width for _ in range(height)]
    visited[start[0]][start[1]] = True

    while queue:
        y, x = queue.popleft()
        if (y, x) == goal:
            print("Yes")
            return

        for dy, dx in DIRECTIONS:
            ny = y + dy
            nx = x + dx

            if ny < 0 or ny >= height:
                continue
            if nx < 0 or nx >= width:
                continue
            if grid[ny][nx] == "#":
                continue
            if visited[ny][nx]:
                continue

            visited[ny][nx] = True
            queue.append((ny, nx))

    print("No")


if __name__ == "__main__":
    main()
